File: express/src/models/crud/place.ts
import {
    PlaceDB,
    PlaceCreate,
    PlacePut,
    PlaceRead,
    PlacePost,
    PlaceUpdate,
    UserRead,
} from "../schemas";
import { PlaceModel, PlaceDocument } from "../collections";
import { storage } from "../../lib/clients";
import { Crud } from "./base";
import { ApiError, HttpStatus, FilterQuery } from "../../types";

export class CrudPlace extends Crud<
    PlaceDB,
    PlaceDocument,
    PlaceRead,
    PlaceCreate,
    PlacePost,
    PlaceUpdate,
    PlacePut
> {
    constructor() {
        super(PlaceModel);
    }

    protected defaultProjection = { __v: 0 } as const;

    public safeCheck(
        user: UserRead,
        data: PlaceDocument | PlacePost | PlaceCreate
    ): void {
        if (!user) {
            throw new ApiError(HttpStatus.UNAUTHORIZED, "Not Authenticated");
        }
        if (user.isAdmin) return;
        const creatorId = "creatorId" in data ? data.creatorId : undefined;
        if (creatorId && creatorId !== user.id) {
            throw new ApiError(
                HttpStatus.UNAUTHORIZED,
                `Cannot set creatorId to ${creatorId}. Access denied`
            );
        }
    }

    public safeFilter(user: UserRead, filterQuery: FilterQuery): FilterQuery {
        const { sort, filters, pagination } = filterQuery;
        if (filters) {
            filters.creatorId = { $eq: user.id };
        }
        return { sort, filters, pagination };
    }

    public async jsonifyBatch(docs: PlaceDocument[]): Promise<PlaceRead[]> {
        const placesPromises = docs.map(async (doc) => {
            let obj = this.serializeDocument(doc);
            if (obj.imageUrl) {
                obj.imageUrl = await storage.getSignedUrl(obj.imageUrl);
            }
            return obj;
        });
        return await Promise.all(placesPromises);
    }

    public async create(form: PlacePost): Promise<PlaceRead> {
        const imageUrl = await storage.uploadFile(form.image || null);
        const { image, ...body } = form;
        const data = { ...body, imageUrl };
        const doc = await this.createDocument(data);
        return this.jsonfify(doc);
    }

    public async update(
        obj: PlaceDocument,
        form: PlacePut
    ): Promise<PlaceRead> {
        const doc = await this.updateDocument(obj, form);
        return await this.jsonfify(doc);
    }

    public async deleteCleanup(document: PlaceDocument): Promise<void> {
        if (document.imageUrl) {
            storage.deleteFile(document.imageUrl);
        }
    }
}

export const crudPlace = new CrudPlace();

File: backend-fastapi/models/crud/place.py
from http import HTTPStatus
from typing import cast

from beanie import PydanticObjectId

from lib.clients import cloud_storage
from models.collections.place import Place
from models.crud.base import CrudBase, CrudEvent
from models.fields import (
    PlaceSearchableFields,
    PlaceSelectableFields,
    PlaceSortableFields,
)
from models.schemas import (
    PlaceCreateSchema,
    PlaceFiltersSchema,
    PlaceFindQuery,
    PlacePostSchema,
    PlacePutSchema,
    PlaceReadSchema,
    PlaceUpdateSchema,
    UserReadSchema,
)
from types_ import ApiError, Filter
from worker.tasks import place_embeddding


class CrudPlace(
    CrudBase[
        Place,
        PlaceReadSchema,
        PlaceSortableFields,
        PlaceSelectableFields,
        PlaceSearchableFields,
        PlaceFiltersSchema,
        PlaceCreateSchema,
        PlacePostSchema,
        PlaceUpdateSchema,
        PlacePutSchema,
    ]
):
    FILTER_FIELD_MAPPING: dict[str, str] = dict(
        locationLat="location.lat", locationLng="location.lng"
    )

    def auth_check(
        self,
        user: UserReadSchema,
        data: Place | PlacePostSchema | PlacePutSchema,
        event: CrudEvent,
    ) -> None:
        if user is None:
            raise ApiError(HTTPStatus.UNAUTHORIZED, "Not Authenticated")
        if user.isAdmin:
            return

        creatorId: PydanticObjectId | None = getattr(data, "creatorId", None)
        if creatorId and str(creatorId) != str(user.id):
            raise ApiError(
                HTTPStatus.UNAUTHORIZED,
                f"Access denied to creator {creatorId}",
            )

    def add_ownership_filters(
        self, user: UserReadSchema, query: PlaceFindQuery
    ) -> PlaceFindQuery:
        ownership_filters = [Filter(op="eq", val=user.id)]

        if query.filters is None:
            query.filters = {}

        creatorId_filters = query.filters.get("creatorId", [])
        creatorId_filters.extend(ownership_filters)
        query.filters["creatorId"] = creatorId_filters
        return query

    async def _post_process_dict(self, item: dict) -> dict:
        item = await super()._post_process_dict(item)
        item.pop("embedding", None)
        image_url: str | None = item.get("imageUrl", None)
        if image_url is not None:
            item["imageUrl"] = cloud_storage.get_signed_url(image_url)
        return item

    async def create(self, form: PlacePostSchema) -> PlaceReadSchema:
        data = form.model_dump()
        lat = data.pop("lat")
        lng = data.pop("lng")
        data["location"] = dict(lat=lat, lng=lng)
        image = data.pop("image", None)
        data["imageUrl"] = cloud_storage.upload_file(image)
        create_form = PlaceCreateSchema(**data)
        document = await self.create_document(create_form)
        place_embeddding(document.id)
        result = await self.post_process(document)
        return cast(PlaceReadSchema, result)

    async def update(self, document: Place, form: PlacePutSchema) -> PlaceReadSchema:
        j = form.model_dump(exclude_none=True, exclude_unset=True)
        update = self.update_schema(**j)
        document = await self.update_document(document, update)

        # Trigger new embedding if title or description changed
        description_changed = (
            form.description and form.description != document.description
        )
        title_changed = form.title and form.title != document.title
        if description_changed or title_changed:
            place_embeddding(document.id)

        result = await self.post_process(document)
        return cast(PlaceReadSchema, result)

    async def delete_cleanup(self, document: Place):
        if document.imageUrl:
            cloud_storage.delete_file(document.imageUrl)


crud_place = CrudPlace()
